config/configtls/clientcasfilereloader.go

```diff
@@ -0,0 +1,148 @@
+	watcher         *fsnotify.Watcher
+	shutdownCH      chan bool
+}
+
+type clientCAsFileLoader interface {
+	loadClientCAFile() (*x509.CertPool, error)
+}
+
+func newClientCAsReloader(clientCAsFile string, loader clientCAsFileLoader) (*clientCAsFileReloader, error) {
+	certPool, err := loader.loadClientCAFile()
+	if err != nil {
+		return nil, fmt.Errorf("failed to load TLS config: failed to load client CA CertPool: %w", err)
+	}
+
+	reloader := &clientCAsFileReloader{
+		ClientCAsFile: clientCAsFile,
+		certPool:      certPool,
+		loader:        loader,
+		shutdownCH:    nil,
+		watcher:       nil,
+	}
+
+	return reloader, nil
+}
+
+func (r *clientCAsFileReloader) GetClientConfig(original *tls.Config) (*tls.Config, error) {
+	r.lock.RLock()
+	defer r.lock.RUnlock()
+	return &tls.Config{
+		RootCAs:              original.RootCAs,
+		GetCertificate:       original.GetCertificate,
+		GetClientCertificate: original.GetClientCertificate,
+		MinVersion:           original.MinVersion,
+		MaxVersion:           original.MaxVersion,
+		NextProtos:           original.NextProtos,
+		ClientCAs:            r.certPool,
+		ClientAuth:           tls.RequireAndVerifyClientCert,
+	}, nil
+}
+
+func (r *clientCAsFileReloader) reload() {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	certPool, err := r.loader.loadClientCAFile()
+	if err != nil {
+		r.certPool = nil
+		r.lastReloadError = fmt.Errorf("%w", err)
+	} else {
+		r.certPool = certPool
+		r.lastReloadError = nil
+	}
+}
+
+func (r *clientCAsFileReloader) StartWatching() error {
+	if r.shutdownCH != nil {
+		return fmt.Errorf("client CA file watcher already started")
+	}
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("failed to create watcher to reload client CA CertPool: %w", err)
+	}
+	r.watcher = watcher
+
+	err = watcher.Add(r.ClientCAsFile)
+	if err != nil {
+		return fmt.Errorf("failed to add client CA file to watcher: %w", err)
+	}
+
+	go r.handleWatcherEvents()
+
+	return nil
+}
+
+func (r *clientCAsFileReloader) handleWatcherEvents() {
+	defer r.watcher.Close()
+	r.shutdownCH = make(chan bool)
+	for {
+		select {
+		case _, ok := <-r.shutdownCH:
+			_ = ok
+			r.shutdownCH = nil
+			return
+		case event, ok := <-r.watcher.Events:
+			if !ok {
+				continue
+			}
+			// NOTE: k8s configmaps uses symlinks, we need this workaround.
+			// original configmap file is removed.
+			// SEE: https://martensson.io/go-fsnotify-and-kubernetes-configmaps/
+			if event.Op == fsnotify.Remove || event.Op == fsnotify.Chmod {
+				// remove the watcher since the file is removed
+				if err := r.watcher.Remove(event.Name); err != nil {
+					r.lastReloadError = err
+				}
+				// add a new watcher pointing to the new symlink/file
+				if err := r.watcher.Add(r.ClientCAsFile); err != nil {
+					r.lastReloadError = err
+				}
+				r.reload()
+			}
+			if event.Op == fsnotify.Write {
+				r.reload()
+			}
+		}
+	}
+}
+
+func (r *clientCAsFileReloader) Shutdown() error {
```

Comment:
Make it private?

Reply:
I removed it as it was not used. Actually this bring a question, what would be a good place to call this method and allow file watcher to close gracefully?